### DISCUSSION TITLE
Download single Asset from private repository in c#

### DISCUSSION THREAD

**Discussion by IT-Emilio** [community/community#110870] [Programming Help]
Hi everyone,

I am building a C# application that should be able to download an asset from a specific release of my private repository. I have created a personal token to access the files and have given it the necessary permissions. According to the Git documentation, I should use a URL like: “https://github.com/{user}/{repo}/releases/download/{tag}/{asset-name}”

However, no matter what I try, I keep receiving a “404” response from the server. This is strange because when I paste the same URL into the browser, the resource downloads correctly.

Here is a code example of what I did on my application, but I have tried several options and none of these seems to work:

```
public static async Task DownloadSingleAssetFile(string Token, string DesideredVersion, string ASSETNAME, string DestinationPath)
{
    var url = $"https://github.com/MYNAME/REPONAME/releases/download/{DesideredVersion}/{ASSETNAME}";
    using (var client = new System.Net.Http.HttpClient())
    {
        var credentials = string.Format(System.Globalization.CultureInfo.InvariantCulture, "{0}:", Token);
        credentials = Convert.ToBase64String(System.Text.Encoding.ASCII.GetBytes(credentials));
        client.DefaultRequestHeaders.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Basic", credentials);
        var contents = client.GetByteArrayAsync(url).Result;
        System.IO.File.WriteAllBytes(DestinationPath, contents);
    }
}
```

So, I’m wondering if it’s even possible to do what I’m trying to achieve. It seems odd that it’s impossible to download a single asset from a repository. I’ve found other people online who seem to be having the same problem, but I haven’t found any solutions.

Interestingly, when I try to use a URL like: “https://github.com/{user}/{repo}/archive/refs/tags/{release-tag}.zip” with the same code, it works correctly.

I hope someone can help me with this. Thanks for your time!

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by asiryan** [comment]
Hello! You need to use your file's URL like this: https://raw.githubusercontent.com/[proj]/[repo]/main/[file]

**Reply by IT-Emilio** [reply]
The support team from Git got back to me. They suggested using a URL like: https://api.github.com/repos/{user}/{repo}/releases/assets/{asset-name}

**Comment by jfullstackdev** [comment]
I asked Copilot to check and correct it, again, this may not be correct but it points out valid points that you can use or may realize

answer from Copilot :

This C# method is designed to download a specific asset file from a GitHub repository's release. Here's a breakdown of what it does:

1. It constructs a URL to the desired asset file in a specific release of a GitHub repository.
2. It creates a new `HttpClient` instance to make HTTP requests.
3. It prepares the authorization header by encoding the provided GitHub token in Base64 and setting it as the `Authorization` header in the `HttpClient`.
4. It sends a GET request to the constructed URL to download the asset file.
5. It writes the downloaded file to the specified destination path.

The method is marked as `async` but it's not correctly using `async/await` pattern. The line `var contents = client.GetByteArrayAsync(url).Result;` is synchronously blocking on the result of an asynchronous operation, which can lead to deadlocks in certain contexts (like in UI applications).

A better approach would be to `await` the `GetByteArrayAsync` method:

```csharp
public static async Task DownloadSingleAssetFile(string Token, string DesideredVersion, string ASSETNAME, string DestinationPath)
{
    var url = $"https://github.com/MYNAME/REPONAME/releases/download/{DesideredVersion}/{ASSETNAME}";
    using (var client = new System.Net.Http.HttpClient())
    {
        var credentials = string.Format(System.Globalization.CultureInfo.InvariantCulture, "{0}:", Token);
        credentials = Convert.ToBase64String(System.Text.Encoding.ASCII.GetBytes(credentials));
        client.DefaultRequestHeaders.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Basic", credentials);
        var contents = await client.GetByteArrayAsync(url);
        System.IO.File.WriteAllBytes(DestinationPath, contents);
    }
}
```

Please replace `MYNAME` and `REPONAME` with your actual GitHub username and repository name. Also, ensure that the GitHub token has the appropriate permissions to access the repository and its releases.

**Comment by asiryan** [comment]
@IT-Emilio , @jdevstatic 

I download GitHub files using Octokit.NET and HttpClient:

```c#
var gitHubClient = new GitHubClient(new ProductHeaderValue(_settings.AppName));
var gitHubCredentials = new Credentials(token: _settings.GitHubToken);
gitHubClient.Credentials = gitHubCredentials;

// get raw file url
var repositoryContents =
    await gitHubClient.Repository.Content.GetAllContentsByRef(
    environmentSettings.Owner,
    environmentSettings.Repository,
    environmentSettings.Branch);

var firstOrDefault = repositoryContents.FirstOrDefault(c => c.Path == path);
var downloadUrl = firstOrDefault?.DownloadUrl;

// download raw file content
using var httpClient = new HttpClient();
var credentials = string.Format(System.Globalization.CultureInfo.InvariantCulture, "{0}:", _settings.GitHubToken);
credentials = Convert.ToBase64String(System.Text.Encoding.ASCII.GetBytes(credentials));
httpClient.DefaultRequestHeaders.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Basic", credentials);

return await httpClient.GetAsync(downloadUrl, cancellationToken)
    .ConfigureAwait(false);
```

**Reply by jfullstackdev** [reply]
is that your latest code ? yes consuming GitHub API is usually done using Octokit, what's the next problem now ?

**Reply by IT-Emilio** [reply]
I discovered that the issue was with the URL. By using CURL from the command line, I was finally able to download the individual asset. So in c#, I decided to download the asset using an HTTP request, without employing the Octokit package. Here’s how I did it

```csharp //
url = $"https://api.github.com/repos/{user}/{repository}/releases/assets/{asset_id}";
using HttpClient dwnclient = new HttpClient();
dwnclient.DefaultRequestHeaders.Add("Authorization", $"Bearer {Token}");
dwnclient.DefaultRequestHeaders.Add("User-Agent", "request");
dwnclient.DefaultRequestHeaders.Add("Accept", "application/octet-stream");

if (!string.IsNullOrEmpty(url))
{
    var fileResponse = await dwnclient.GetAsync(url);
    if (fileResponse.IsSuccessStatusCode)
    {
        DestinationPath = Path.Combine(DestinationPath, AssetName);
        using FileStream fs = new FileStream(DestinationPath, System.IO.FileMode.Create);
        await fileResponse.Content.CopyToAsync(fs);
        return new Result(true, $"The asset '{AssetName}' from version '{desideredVersionTag}' has been successfully downloaded.");
    }
    else
    {
        return new Result(false, $"An error occurred while downloading the asset '{AssetName}': {fileResponse.StatusCode}");
    }
}
else
{
    return new Result(false, $"The asset '{AssetName}' is not present in version '{desideredVersionTag}'.");
}
```

Thanks a lot for your time anyway. 😊

**Reply by asiryan** [reply]
@jdevstatic yes, it works!
@IT-Emilio thanks for your code too

**Reply by leedavi** [reply]
Yes, Thankyou, it's been driving me mad all day.
